include/revng/Pipeline/ContainerInspector.h

```diff
@@ -0,0 +1,122 @@
+#pragma once
+
+//
+// This file is distributed under the MIT License. See LICENSE.md for details.
+//
+
+#include "llvm/ADT/STLExtras.h"
+#include "llvm/ADT/SmallVector.h"
+#include "llvm/Support/ManagedStatic.h"
+
+#include "revng/Pipeline/ContainerSet.h"
+#include "revng/Pipeline/Context.h"
+#include "revng/Pipeline/Target.h"
+
+namespace Pipeline {
+
+/// If a downstream user of reveng wishes to extend the enumeration and
+/// removal rule from a container it can do so by extending this class,
+/// specialized with the type of the container in question.
+///
+/// the enumerate and removeTarget methods must be implemented with the
+/// appropriate rules
+///
+///
+template<typename Container>
+class ContainerInspector {
```

Comment:
OK, let's think about this one last time.

In principle, I'd be happy if `Kind` had `ContainerInspector`'s features (i.e., `enumerate` and `remove`), but, as you suggest, this would needlessly couple `Kind`s and `Container`s.
On the other hand we end up wanting classes such as `KindInspector` which inherit from both `Kind` and `ContainerInspector`.

At this point, can't we have a `KindForContainer` that replaces `KindInspector` in a non-LLVM specific fashion, covers 90% of the use cases but doesn't force us to couple all kinds to containers?